core/src/main/scala/org/apache/spark/rdd/CoalescedRDD.scala

```diff
@@ -273,7 +273,8 @@ private class PartitionCoalescer(maxPartitions: Int, prev: RDD[_], balanceSlack:
       groupArr += pgroup
       groupHash.getOrElseUpdate(nxt_replica, ArrayBuffer()) += pgroup
       var tries = 0
-      while (!addPartToPGroup(nxt_part, pgroup) && tries < targetLen) { // ensure at least one part
+      // ensure at least one part
+      while (pgroup.arr.isEmpty && !addPartToPGroup(nxt_part, pgroup) && tries < targetLen) {
```

Comment:
Is pgroup.arr an array ? (Sorry, on a phone). Use .length == 0 to avoid an implicit conversion if speed is essential